index: Fix volk_get_index

jdemel opened the pull request:
This function results in an infinite loop on Debian 11 for some impls.
This is a first step to fix it.

@mbr0wn , @marcusmueller does that fix the reported issue? If you could point me to a CI script for Debian 11 that we can integrate here, that'd be very helpful.

Fix #516

EDIT
This is a minimal Debian 11 Container to test things.
```docker
FROM debian:bullseye

RUN apt-get update \
        && apt-get install -y \
        build-essential git cmake python3-mako python3-distutils liborc-dev
```
[GNU Radio CI container for Debian 11](https://hub.docker.com/layers/gnuradio/ci/debian-11-3.10/images/sha256-d3de16fb91a50ee2472d35bd1fb27e18715e4dba69098bf3b99e0109aaf0c62c?context=explore) where `volk_profile` hangs on `volk_32f_8u_polarbutterflypuppet_32f`.
The corresponding Dockerfile: [Dockerfile](https://github.com/gnuradio/gnuradio-docker/blob/master/ci/ci-debian-11-3.10/Dockerfile).

jdemel committed:
index: Fix volk_get_index

This function results in an infinite loop on Debian 11 for some impls.
This is a first step to fix it.

Fix #516

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

marcusmueller commented:
Don't we need to handle -1 in the calling functions to avoid calling into random memory?

michaelld commented:
> Don't we need to handle -1 in the calling functions to avoid calling into random memory?

+1 from me on this: looks like that's just in tmpl/volk.tmpl.c : 191

michaelld commented on lib/volk_rank_archs.c:
```diff
@@ -34,13 +34,16 @@ int volk_get_index(const char* impl_names[], // list of implementations by name
     unsigned int i;
     for (i = 0; i < n_impls; i++) {
-        if (!strncmp(impl_names[i], impl_name, 20)) {
+        if (!strncmp(impl_names[i], impl_name, 42)) {
```
since `strncmp` returns `int` of "less than, equal to, or greater than zero" can we go with `strncmp () == 0` here? Compilers will know how to optimize the code here, and this avoids any possible issues with binary not (`!`) and negative integers. Yes this code should work fine, but it's just as easy to compare with 0.

jdemel replied:
I really hope it is `!= 0`. Fixed it.

michaelld commented on lib/volk_rank_archs.c:
```diff
@@ -34,13 +34,16 @@ int volk_get_index(const char* impl_names[], // list of implementations by name
     // TODO return -1;
     // something terrible should happen here
     fprintf(stderr, "Volk warning: no arch found, returning generic impl\n");
+    if (strncmp(impl_name, "generic", 20)) {
```
can we use 7 or 8 instead of 20? "generic" is 7 bytes long, so 20 way more than necessary ...

michaelld replied:
IDK whether we use 7 or 8 bytes ... would have to test `strncmp` to see ...

michaelld replied:
Also: If we're adding in this check, then let's save the result of this `strncmp` and return it instead of the recursive call to this function. Once through and done.

jdemel replied:
If we change the recursive structure, we'd need to change the whole function quite significantly.
Right now, we follow a 3 step process:
1. Search for the user requested impl
2. If user request cannot be fulfilled: Search for generic impl.
3. New third step: Give up and return `-1`.
I'm actually not entirely sure, if the term `generic` is the only thing we ever search for. It might be smth like `vector_generic` as well. However, not during the second pass.

Right now, it feels like we actually need 2 functions. One with the loop and one to handle all the different error conditions.

jdemel committed:
index: Restructure volk_get_index

Instead of recursively calling `volk_get_index`, a new function is
introduced to walk through the list of impls. Then, we introduce a three
step process:
1. Search for the requested impl. Return index if found.
2. Search for the generic impl. Return index if found.
3. Return `-1` and put the burden on the caller.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel commented:
I fixed the "file local comments". Thus, we need to discuss how to deal with the `-1` return value.
Why did we avoid to deal with it until now? We ended up in an infinite loop and thus terminated useful execution at this point.
Considering this, the actual bug report should be fixed now.

`volk_get_index` is used in 2 functions:
1. `lib/volk_rank_archs.c:volk_rank_archs`: return values propagate here.
2. `tmpl/volk.tmpl.c:186`, i.e. in all `_manual` kernel calls.

`volk_rank_archs` is used in "one" place.
- `tmpl/volk.tmpl.c:151-152`: kernel init. Thus, the first call to any kernel.

Both functions in `tmpl/volk.tmpl.c` return void. We can't signal errors that way. We'd have to change the API.
We go back and forth between signed and unsigned variables in these functions. But that's a separate issue.

We could treat `-1` in both functions as: use the first impl in the list and hope it is supported on your system.
I'd really like to raise a `NotImplementedError` but that's obviously not an option either.

I know this is an unsatisfactory but we could ignore this problem for now and add it to the list of issues to fix for our next major release. Would that be a viable option?

jdemel committed:
index: Revert to !strncmp

This fix is a bit of a guess. aarch64, MacOS, and Win CI tests all fail
after I changed the comparison from `!` to `!= 0`.

Signed-off-by: Johannes Demel <demel@uni-bremen.de>

jdemel commented:
I reverted the change where `!` is replaced with `!= 0`. It caused issues with most systems except Ubuntu x86. I rebased to the latest `main` as well.

michaelld reviewed (approved):
LGTM; this change will be an API change (returning -1), so do we want to just have it in a 3.0-beta branch & not related to 2.X at all?

michaelld commented:
> I reverted the change where `!` is replaced with `!= 0`. It caused issues with most systems except Ubuntu x86. I rebased to the latest `main` as well.

interesting ... I wonder why? Anyway, `!` is fine

michaelld commented:
This PR fixes the direct issue, which is important. It does change the API unfortunately by now returning -1 where it didn't before. I guess as an interim we could instead have it return the first entry (if there is one) & print a warning noting as such. Of course if there is no entry then we'd want to return -1 ... so, does this help? Hmmm ...

jdemel commented:
> This PR fixes the direct issue, which is important. It does change the API unfortunately by now returning -1 where it didn't before. I guess as an interim we could instead have it return the first entry (if there is one) & print a warning noting as such. Of course if there is no entry then we'd want to return -1 ... so, does this help? Hmmm ...

It sounds like every solution somehow breaks the API. However, at the moment the broken API won't work at all. After this change, it will. If we'd do another 2.x release, I'd argue this fix should go in there.
I added a message "Volk ERROR ...". If we'd `return 0;`, we'd most probably cause an "illegal instruction error". 

Besides, it might be time to add a Debian test. However, [Debian 'bullseye' ships GCC 10](https://packages.debian.org/en/bullseye/devel/gcc). We already have a test with GCC 10. The question would be: Is this distro specific?

marcusmueller commented on lib/volk_rank_archs.c:
```diff
@@ -27,21 +27,43 @@
+    } else {
+        fprintf(stderr, "Volk ERROR: no arch found. generic impl missing!\n");
+    }
+    return idx;
```
we're just bubbling up the -1 here in case of error, but we need to handle that. Otherwise, we'll jump into some random memory location when the consumer program actually calls the kernel. That's worse than an infinite loop.

So, this is half the fix so far.

marcusmueller replied:
at the very least, `__init_${kern.name}` needs to check for -1 and `abort` in that case. Failing early instead of wandering into random memory, where we become a stability or security issue, is necessary. Otherwise, we're just making bad code worse!

marcusmueller replied:
also note that `__init_..` casts the `int` that `volk_rank_archs` returns into a `size_t`, which is unsigned, and potentially of a different size!

marcusmueller commented:
By the way, how do we figure out what breaks finding that `generic` implementation? even if this PR was perfect in every way, it would *not* solve the underlying issue that leads to https://github.com/gnuradio/volk/issues/516 . And maybe fixing that would completely rewrite this part of code, anyway...

jdemel commented:
Updated the initial PR description.
I started some tests with:
```docker
FROM debian:bullseye

RUN apt-get update \
        && apt-get install -y \
        build-essential git cmake python3-mako python3-distutils liborc-dev
```
I couldn't find `liborc-dev` and `python3-distutils` in the GNU Radio CI Docker file.
`liborc-dev` is essentially a dependency and should be available anyways.
`python3-distutils` is only a build dependency. So it won't be installed with a binary.

Besides, I can't reproduce the issue in my minimal container with the current `main`. Debian 11 ships VOLK 2.4.1 which I didn't test yet.

marcusmueller commented:
> liborc-dev is essentially a dependency and should be available anyways.

liborc-**dev** should only be a build-time dependency, right?

jdemel commented:
I converted this PR to a draft. At the moment I doubt it really fixes the issue it is supposed to fix. We need a better understanding of the issue here.

jdemel commented:
This draft PR was introduced because of a bug that turned out to be outside of VOLK. I'm closing this PR now.
